### PR TITLE
Propagate context to child function

### PR DESCRIPTION
**Pull request by A-kanksh-a**
## Change Overview
This PR is to refactor code and pass the same context object if it’s available in parent call while performing `Create`, `List`, `Get` operation. 
Referred PR [https://github.com/kanisterio/kanister/pull/726](https://github.com/kanisterio/kanister/pull/726)

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [x] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues

- K10-5149

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/controller/controller_test.go [hunk 1]
```diff
@@ -437,7 +437,7 @@ func (s *ControllerSuite) TestExecActionSet(c *C) {
 
 func (s *ControllerSuite) TestRuntimeObjEventLogs(c *C) {
 	c.Skip("This may not work in MiniKube")
-	ctx := context.TODO()
+	contxt := context.TODO()
```
Comment (viveksinghggits) [inline pkg/controller/controller_test.go:440]:
```suggestion
	ctx := context.TODO()
```

Comment (A-kanksh-a) [reply pkg/controller/controller_test.go:440]: I don't think we should do that, `ctx` is being initialized later in the function with `context.Background` in line no 468.

Comment (viveksinghggits) [reply pkg/controller/controller_test.go:440]: so why don't we use this one there as well.

File: pkg/kube/pod_writer_test.go [hunk 2]
```diff
@@ -37,14 +37,15 @@ var _ = Suite(&PodWriteSuite{})
 
 func (p *PodWriteSuite) SetUpSuite(c *C) {
 	var err error
+	contxt := context.TODO()
```
Comment (A-kanksh-a) [inline pkg/kube/pod_writer_test.go:40]: @viveksinghggits here I have used `contxt `as we are initializing `ctx` with a timeout at line 65

File: pkg/param/param_test.go [hunk 3]
```diff
@@ -359,6 +359,7 @@ func (s *ParamsSuite) testNewTemplateParams(ctx context.Context, c *C, dynCli dy
 	c.Assert(err, IsNil)
 	c.Assert(cm, NotNil)
 
+	contxt := context.TODO()
```
Comment (A-kanksh-a) [inline pkg/param/param_test.go:362]: @viveksinghggits here `ctx` is already initialized as a parameter.

File: pkg/kube/unstructured.go [hunk 4]
```diff
@@ -35,10 +35,11 @@ func FetchUnstructuredObject(resource schema.GroupVersionResource, namespace, na
 // FetchUnstructuredObjectWithCli returns the referenced API object as a map[string]interface{} using the specified CLI
 // TODO: deprecate `FetchUnstructuredObject`
 func FetchUnstructuredObjectWithCli(cli dynamic.Interface, resource schema.GroupVersionResource, namespace, name string) (runtime.Unstructured, error) {
+	ctx := context.TODO()
```
Comment (pavannd1) [inline pkg/kube/unstructured.go:38]: Pass `ctx` in from `param.New`

File: pkg/kube/unstructured.go [hunk 5]
```diff
@@ -53,10 +54,11 @@ func ListUnstructuredObject(resource schema.GroupVersionResource, namespace stri
 // ListUnstructuredObjectWithCli returns the referenced API objects as a map[string]interface{} using the specified CLI
 // TODO: deprecate `ListUnstructuredObject`
 func ListUnstructuredObjectWithCli(cli dynamic.Interface, resource schema.GroupVersionResource, namespace string) (runtime.Unstructured, error) {
+	ctx := context.TODO()
```
Comment (pavannd1) [inline pkg/kube/unstructured.go:57]:
```suggestion
	ctx := context.Background()
```

File: pkg/param/param_test.go (hunk 3 above)
Comment (pavannd1) [reply pkg/param/param_test.go:362]: Use `ctx`
```suggestion
```

File: pkg/param/param_test.go [hunk 6]
```diff
@@ -387,21 +388,21 @@ func (s *ParamsSuite) testNewTemplateParams(ctx context.Context, c *C, dynCli dy
 			},
 		},
 	}
-	_, err = s.cli.CoreV1().Secrets(s.namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+	_, err = s.cli.CoreV1().Secrets(s.namespace).Create(contxt, secret, metav1.CreateOptions{})
```
Comment (pavannd1) [inline pkg/param/param_test.go:391]:
```suggestion
	_, err = s.cli.CoreV1().Secrets(s.namespace).Create(ctx, secret, metav1.CreateOptions{})
```

File: pkg/param/param_test.go [hunk 7]
```diff
@@ -387,21 +388,21 @@ func (s *ParamsSuite) testNewTemplateParams(ctx context.Context, c *C, dynCli dy
...
 	c.Assert(err, IsNil)
 	defer func() {
 		_ = s.cli.CoreV1().Secrets(s.namespace).Delete(context.TODO(), "secret-name", metav1.DeleteOptions{})
 	}()
 
-	_, err = s.cli.CoreV1().Secrets(s.namespace).Get(context.TODO(), "secret-name", metav1.GetOptions{})
+	_, err = s.cli.CoreV1().Secrets(s.namespace).Get(contxt, "secret-name", metav1.GetOptions{})
```
Comment (pavannd1) [inline pkg/param/param_test.go:397]:
```suggestion
	_, err = s.cli.CoreV1().Secrets(s.namespace).Get(ctx, "secret-name", metav1.GetOptions{})
```

File: pkg/param/param_test.go [hunk 8]
```diff
@@ -387,21 +388,21 @@ func (s *ParamsSuite) testNewTemplateParams(ctx context.Context, c *C, dynCli dy
...
 	c.Assert(err, IsNil)
 
 	osCli := osfake.NewSimpleClientset()
 
 	crCli := crfake.NewSimpleClientset()
-	_, err = crCli.CrV1alpha1().Profiles(s.namespace).Create(context.TODO(), prof, metav1.CreateOptions{})
+	_, err = crCli.CrV1alpha1().Profiles(s.namespace).Create(contxt, prof, metav1.CreateOptions{})
```
Comment (pavannd1) [inline pkg/param/param_test.go:403]:
```suggestion
	_, err = crCli.CrV1alpha1().Profiles(s.namespace).Create(ctx, prof, metav1.CreateOptions{})
```

File: pkg/param/param_test.go [hunk 9]
```diff
@@ -387,21 +388,21 @@ func (s *ParamsSuite) testNewTemplateParams(ctx context.Context, c *C, dynCli dy
...
 	c.Assert(err, IsNil)
-	_, err = crCli.CrV1alpha1().Profiles(s.namespace).Get(context.TODO(), "profName", metav1.GetOptions{})
+	_, err = crCli.CrV1alpha1().Profiles(s.namespace).Get(contxt, "profName", metav1.GetOptions{})
```
Comment (pavannd1) [inline pkg/param/param_test.go:405]:
```suggestion
	_, err = crCli.CrV1alpha1().Profiles(s.namespace).Get(ctx, "profName", metav1.GetOptions{})
```

File: pkg/testing/integration_test.go [hunk 10]
```diff
@@ -304,15 +304,16 @@ func newActionSet(bpName, profile, profileNs string, object crv1alpha1.ObjectRef
 }
 
 func (s *IntegrationSuite) createProfile(c *C) string {
-	secret, err := s.cli.CoreV1().Secrets(kontroller.namespace).Create(context.TODO(), s.profile.secret, metav1.CreateOptions{})
+	ctx := context.TODO()
```
Comment (pavannd1) [inline pkg/testing/integration_test.go:307]: Pass `ctx` in from `TestRun`

File: pkg/testutil/testutil_test.go [hunk 11]
```diff
@@ -42,25 +42,25 @@ func (s *TestUtilSuite) TestDeployment(c *C) {
 	c.Assert(err, IsNil)
 	crCli, err := crclientv1alpha1.NewForConfig(config)
 	c.Assert(err, IsNil)
-
+	ctx := context.TODO()
```
Comment (pavannd1) [inline pkg/testutil/testutil_test.go:45]:
```suggestion
	ctx := context.Background()
```

File: pkg/testutil/testutil_test.go [hunk 12]
```diff
@@ -84,17 +84,18 @@ func (s *TestUtilSuite) TestDeployment(c *C) {
...
 
 	cm := NewTestConfigMap()
-	cm, err = cli.CoreV1().ConfigMaps(ns.GetName()).Create(context.TODO(), cm, metav1.CreateOptions{})
+	cm, err = cli.CoreV1().ConfigMaps(ns.GetName()).Create(ctx, cm, metav1.CreateOptions{})
 	c.Assert(err, IsNil)
 	err = cli.CoreV1().ConfigMaps(ns.GetName()).Delete(context.TODO(), cm.GetName(), metav1.DeleteOptions{})
 	c.Assert(err, IsNil)
 }
 
 func testCRs(c *C, cli crclientv1alpha1.CrV1alpha1Interface, namespace, poKind, poName string) {
```
Comment (pavannd1) [inline pkg/testutil/testutil_test.go:93]: Pass it in from above
```suggestion
func testCRs(c *C, ctx context.Context, cli crclientv1alpha1.CrV1alpha1Interface, namespace, poKind, poName string) {
```

File: pkg/testutil/testutil_test.go [hunk 13]
```diff
@@ -84,17 +84,18 @@ func (s *TestUtilSuite) TestDeployment(c *C) {
...
 	var err error
+	ctx := context.TODO()
```
Comment (pavannd1) [inline pkg/testutil/testutil_test.go:95]:
```suggestion
```

File: pkg/kube/snapshot/snapshot_alpha.go [hunk 14]
```diff
@@ -62,7 +62,8 @@ func NewSnapshotAlpha(kubeCli kubernetes.Interface, dynCli dynamic.Interface) Sn
 
 // CloneVolumeSnapshotClass creates a copy of the source volume snapshot class
 func (sna *SnapshotAlpha) CloneVolumeSnapshotClass(sourceClassName, targetClassName, newDeletionPolicy string, excludeAnnotations []string) error {
```
[the remark this answers is not in the document]
Comment (PrasadG193) [reply pkg/kube/snapshot/snapshot_alpha.go:64]: Yes, let's refactor and update the interface to pass context as argument @A-kanksh-a

**Review by pavannd1 — APPROVED** [review]
LGTM

**Pull request merged**